Document a procedure for iterative development without needing a full container rebuild

acmorrow opened the pull request:
@penguinland - Please try this out locally and let me know if it works for you.

acmorrow commented:
@penguinland - I just realized that there is an issue with this process for the NVCR containers, and I'll need to give it a little more thought.

acmorrow commented on README.development.md:
````diff
@@ -0,0 +1,153 @@
+
+```
+$ cd ~/opt/src/viam-mlmodelservice-triton
+$ cmake -S . -B build -G Ninja -DVIAM_MLMODELSERVICE_TRITON_TRITONSERVER_ROOT=/opt/tritonserver/tritonserver -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_PREFIX_PATH=/opt/viam
````
This command line should include probably include generating the compile commands db.

penguinland replied:
There's no way you expect me to remember this line and type it out every time, and I suspect you don't intend for me to need to open this file to copy and paste it every time. How do you actually invoke this? 

...I wonder if the answer is "similar to our [discussion last week](https://github.com/viam-modules/viam-mlmodelservice-triton/pull/30#discussion_r1936205532), once this is documented, it can be automated, whether with `make` or something else."

penguinland replied:
When running this command, I get a giant warning that FindBoost is removed. Is that a problem? Do you get that warning, too?

acmorrow replied:
A few things here:
- These commands are lifted directly from the docker files, like: https://github.com/viam-modules/viam-mlmodelservice-triton/blob/fb9a7a3ebc9e0191584471e1671f9319729d0f66/etc/docker/Dockerfile.triton-jetpack-focal#L140-L145
- You don't need to run this every time you (re)build. Just the first time. So, effectively, once per time you create a new `-dev` docker container.
- The command itself it isn't as bad as it looks and can be recovered mostly from first principles:
-- `cmake -S . -B build -G Ninja` is a standard CMake invocation prelude that asks to run cmake  with a context of the current directory, with a build directory target of `build`, and using the Ninja generator.
-- `-DVIAM_MLMODELSERVICE_TRITON_TRITONSERVER_ROOT=/opt/tritonserver/tritonserver` is telling you the path to where the triton server is installed.
-- `-DCMAKE_BUILD_TYPE=RelWithDebInfo` is standard cmake-eze for an optimized build but which retains debug info.
-- `-DCMAKE_PREFIX_PATH=/opt/viam` tells cmake that it should consider `/opt/viam` as a place to search for packages so it can find the installed C++ SDK and, if we built it, grpc.
- Even if you need to copy it in the first time from this document, after that, it will be a ^R away in your shell history inside the container session (especially if you use tmux to keep it alive) if you do find a need to re-run it, such as deleting the `build` directory.
- I think it is likely that this could subsequently be scripted.

acmorrow commented:
@penguinland - OK, this procedure is now working for JP5. I will test it with NVCR containers next. Could you please have a read through and let me know if it makes sense to you? I'm happy to walk you through it in person as well.

penguinland reviewed (commented):
I've run out of disk space again, which means it's time to install the NVMe drive that finally arrived for this machine and start from scratch. Will try doing more steps in an hour...

acmorrow commented on README.development.md:
```diff
@@ -0,0 +1,154 @@
+The first part of the process is to establish a container in which you can do iterative development.
+This is a one time setup cost, so, while there are a few steps, you shouldn't need to repeat them too often.
+
+
```
Done

acmorrow commented on README.development.md:
````diff
@@ -0,0 +1,154 @@
+$ docker build -f ./etc/docker/Dockerfile.triton-jetpack-focal --target build-deps . -t acm-triton-jp5-iterative-dev
+```
+
+If you were going to work on a JP6 board, you would probalby build that more like:
````
Done

acmorrow commented on README.development.md:
````diff
@@ -0,0 +1,154 @@
+```
+
+
+
````
No, no purpose, just editing side effects. I've removed them

acmorrow commented on README.development.md:
````diff
@@ -0,0 +1,154 @@
+If you were going to work on a JP6 board, you would probalby build that more like:
+
+```sh
+$ docker build -f ./etc/docker/Dockerfile.nvcr-triton-containers --target build-deps --build-arg JETPACK=1 . -t acm-triton-jp6-iterative-dev
````
It does. I'll scrub them out.

acmorrow commented on README.development.md:
````diff
@@ -0,0 +1,154 @@
+or
+
+```sh
+$ docker run -id --name acm-triton-jp5-deploy acm-triton-jp6-iterative-deploy
````
Yes, obviously some or all of this is scriptable. Whether `make` is the right vehicle for that or not, I'm not sure. But I'd like to get the procedure documented, and then from there it can be automated.

penguinland commented on README.development.md:
```diff
@@ -0,0 +1,148 @@
+NOTE: I'd like to eliminate this step.
+Potentially, there may be a way to mount the `/opt/viam` (and, when needed for JP5, the `/opt/tritionserver`) directories from the `-dev` container into the `-deploy` container, such that writes in the `-dev` container immediately affect the running `-deploy` container, which would then get saved via the `docker commit`.
+
+3. Overwrite that container image with a snapshot of your deployment container contents
```
:grimacing: I really don't like this approach! It's so easy to mistakenly think you're running the production version when you're actually running a local build that you forgot about last week.

Instead, I'd add a local module and disable the production one. The local module would contain the .sh script which launches the `triton-jp6-deploy` image instead of the `ghcr.io/...` one. That way, any docker image named `ghcr.io/...` comes from the GitHub Container Repository, and any locally-built repo has an obviously-local name.

acmorrow replied:
Sure, that's a fair objection. However, I only promised an effective procedure, not an elegant one :)

I also don't personally and immediately know how to describe the steps to do the local module things vs production. So, if you do have that knowledge and would like to update this document to describe it in a follow-on review, I'll be happy to approve it.

penguinland replied:
I'm working on it now. So far, it's not working because the docker container run by the module's .sh script seems to be a fresh, pristine copy without the extra stuff copied in. I'm relatively new to docker, so might be doing something dumb on that front.

penguinland replied:
Okay, I've got something that works! After the `docker cp` command on line 130, run `docker commit triton-jp6-deploy triton-jp6-iterative-deploy` to turn this updated container back into an image.

Meanwhile, to compile the module itself, run `TAG=triton-jp6-iterative-deploy make -f Makefile.module viam-mlmodelservice-triton.sh` to get the shell script to use the local image name. Then, on app.viam.com, add a local module pointing to that .sh file.

Now, the only docker image that ever gets renamed is the `-iterative-deploy` one, which seems like the right approach.

acmorrow replied:
Great. Let's merge this, and you can amend it with your improved process. I don't think it makes sense for me to write a step by step guide for something you have practice doing.

acmorrow commented on README.development.md:
````diff
@@ -0,0 +1,148 @@
+or
+
+```sh
+$ docker run -id --name triton-jp5-deploy triton-jp6-iterative-deploy
````
Done on next push.

penguinland commented on README.development.md:
````diff
@@ -0,0 +1,148 @@
+```sh
+$ docker exec -it triton-jp5-dev apt-get update
+$ docker exec -it triton-jp5-dev apt-get install tmux
+$ docker exec -it triton-jp5-dev tmux -u -CC new-session -A -s dev
````
Nope, I never use tmux or screen. I just had a terminal SSHed into my Orin Nano, and ran the commands there. Was I supposed to install something else? TBH, I'm hopeful that I don't need to adopt too many new tools just to edit this code.

acmorrow commented on README.development.md:
````diff
@@ -0,0 +1,148 @@
+1. Copy the installation from `/opt/viam` in the development container into the deployment container
+
+```sh
+$ docker cp -a triton-jp5-dev:/opt/viam - | docker cp -a - triton-jp5-deploy:/opt
````
Sure, added.

acmorrow commented:
I don't know why GH has disappeared the review comment regarding `FindBoost`, but it happens in the CI pipeline as well: https://github.com/viam-modules/viam-mlmodelservice-triton/actions/runs/12899862124/job/35969322336#step:6:22560.

So, I'm not too concerned about it. The container base is unlikely to change in a way that causes this warning to become something that needs to be addressed, and the build proceeds just fine.

acmorrow commented:
@penguinland - Updated:
- Made `bash` the primary way to get a session, but retained suggestions for using `tmux` or `screen`.
- Removed the erroneous `-CC` from the `tmux` invocation.
- Added compile commands generation to the CMake invocation to produce the build system
- Added the note about the misleading zero copy statement.
- Clarified where the process can be returned to, to make it easier to understand what the "iteration" pattern looks like (i.e. that you do not need to re-run the scary cmake command each time you make a change).

I think this is ready to merge. Further improvements can be made on this baseline as you experiment with it.

penguinland reviewed (approved):
Well, if you want me to make the changes I want instead of you doing it yourself, I guess there's not much more for me to do besides approve this. I still think the part about overwriting the production image is the wrong approach and most of this document should be a shell script instead of a readme, but it sounds like you're not going to make those changes and it's up to me to make them instead.
